controller/src/metrics.rs

```diff
@@ -0,0 +1,86 @@
+use opentelemetry::{
+    metrics::{Meter, ObserverResult},
+    Key,
+};
+use std::collections::HashMap;
+use std::sync::Arc;
+use std::sync::Mutex;
+
+const HOST_VERSION_KEY: Key = Key::from_static_str("bottlerocket_version");
+const HOST_STATE_KEY: Key = Key::from_static_str("state");
+
+pub struct BrupopControllerMetrics {
+    brupop_shared_hosts_data: Arc<Mutex<BrupopHostsData>>,
+}
+
+pub struct BrupopHostsData {
+    hosts_version_count_map: HashMap<String, u64>,
+    hosts_state_count_map: HashMap<String, u64>,
+}
+
+impl BrupopHostsData {
+    pub fn new(
+        hosts_version_count_map: HashMap<String, u64>,
+        hosts_state_count_map: HashMap<String, u64>,
+    ) -> Self {
+        BrupopHostsData {
+            hosts_version_count_map,
+            hosts_state_count_map,
+        }
+    }
+}
+
+impl Default for BrupopHostsData {
+    fn default() -> Self {
+        let hosts_version_count_map = HashMap::new();
+        let hosts_state_count_map = HashMap::new();
+        BrupopHostsData {
+            hosts_version_count_map,
+            hosts_state_count_map,
+        }
+    }
+}
+impl BrupopControllerMetrics {
+    pub fn new(meter: Meter) -> Self {
+        let brupop_shared_hosts_data = Arc::new(Mutex::new(BrupopHostsData::default()));
+
+        let hosts_data_clone_for_version = Arc::clone(&brupop_shared_hosts_data);
+        let hosts_version_callback = move |res: ObserverResult<u64>| {
+            let data = hosts_data_clone_for_version.lock().unwrap();
+            for (host_version, count) in &data.hosts_version_count_map {
+                let labels = vec![HOST_VERSION_KEY.string(host_version.clone())];
+                res.observe(*count, &labels);
+            }
+        };
+        // Observer for cluster host's bottlerocket version
+        let _brupop_hosts_version = meter
+            .u64_value_observer("brupop_hosts_version", hosts_version_callback)
+            .with_description("Brupop host's bottlerocket version")
+            .init();
+
+        let hosts_data_clone_for_state = Arc::clone(&brupop_shared_hosts_data);
+        let hosts_state_callback = move |res: ObserverResult<u64>| {
+            let data = hosts_data_clone_for_state.lock().unwrap();
```

Comment:
I think we should avoid using `unwrap` : )

Reply:
Do you mean we need write some error handle here?

Reply:
Yeah, we should handle the error here . You can replace `unwrap` with `method.context(error::MyCoolError)?`. Sorry about my unclear expression.

Reply:
I was trying to handle the error here but the use case here is a little bit different.

Here is the definition for the callback clousure:
```
        let hosts_state_callback = move |res: ObserverResult<u64>| {
            let data = hosts_data_clone_for_state.lock().unwrap();
            for (host_state, count) in &data.hosts_state_count_map {
                let labels = vec![HOST_STATE_KEY.string(host_state.clone())];
                res.observe(*count, &labels);
            }
        };
```

And here is the definition for the type of the required callback:
```
pub fn u64_value_observer<T, F>(
    &self,
    name: T,
    callback: F
) -> ValueObserverBuilder<'_, u64> where
    T: Into<String>,
    F: Fn(ObserverResult<u64>) + Send + Sync + 'static,  <-------- callback type
```

I can't return something like `?` with a closure which should return ().

Reply:
If the closure returns (), then we have to decide whether it is better to panic and crash the program, or to log the error and let the program continue as if nothing bad happened. I guess those are the only two choices!

Reply:
It turns out that `Mutex` returns an error on `lock()` [only when it is poisoned](https://doc.rust-lang.org/std/sync/struct.Mutex.html#poisoning), meaning another thread holding the lock has paniced. I think in that scenario, it is acceptable to panic.